docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -182,118 +500,124 @@ pattern-propagators:
   to: $X
 ```
 
-In general, a taint propagator must specify:
-1. A pattern containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**.
-2. The `to` and `from` metavariables. These metavariables should match a **variable** or an **expression**.
-    - The `from` metavariable specifies the entry point of the taint.
-    - The `to` metavariable specifies where the tainted data is propagated to, typically an object or data structure.
+### Propagation without side-effect
 
-:::note
-When the `to` metavariable matches a variable (or in general, an l-value), the taint is propagated _by side-effect_.
-See [Sources by side effect](#sources-by-side-effect) for an explanation of what this means.
-:::
+Taint propagators can be used in very imaginative ways, and in some cases we may not want taint to propagate by side-effect. This can be achieved by disabling `by-side-effect` (enabled be default).
 
-For example, pattern `$S.add($E)` includes two metavariables `$S` and `$E`. Given `from: $E` and `to: $S`, and with `$E` matching `x` and `$S` matching `s`, when `x` is tainted then `s` becomes tainted (by side-effect) with the same taint as `x`.
+For example:
 
-## Sanitizers
+```yaml
+pattern-propagators:
+  - patterns:
+    - pattern: |
+        if something($FROM):
+          ...
+          $TO()
+          ...
+    from: $FROM
+    to: $TO
+    by-side-effect: false
+```
 
-A sanitizer is specified by a pattern, and **any** subexpression that is matched by this pattern is regarded as sanitized.
+The propagator above specifies that inside an `if` block, where the condition is `something($FROM)`, we want to propagate taint from `$FROM` to any function that is being called without arguments, `$TO()`.
 
-:::note
-Given the sanitizer specification below, and a piece of code such as `sanitize(sink(source))`, the call `sink(source)` is **not** reported.
-The reason is that the pattern `sanitize(...)` matches all of `sanitize(sink(source))`, and that makes Semgrep label every subexpression in that piece of code as being sanitized. In particular, `source` is considered to be sanitized!
+<iframe src="https://semgrep.dev/embed/editor?snippet=4bv6x" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-```yaml
-pattern-sanitizers:
-- pattern: sanitize(...)
-```
-:::
+Because we are disabling `by-side-effect`, the `sink` occurence that is inside the `if` block is tainted, but this does not affect the `sink` occurrence outside the `if` block.
```

Comment:
```suggestion
Because the rule disables `by-side-effect`, the `sink` occurrence that is inside the `if` block is tainted, but this does not affect the `sink` occurrence outside the `if` block.
```